### PR TITLE
CA-144169: Removing unneeded XAPI calls for VBD.plug operation.

### PR DESCRIPTION
**Pull request by pritha-srivastava**
CA-144169: Fix to remove XAPI calls to improve time taken by VBD plug.

Signed-off-by: Pritha Srivastava pritha.srivastava@citrix.com

### CODE REVIEW

**Comment by germanop** [comment]
@pritha-srivastava, how much is the performance improvement here? Most of the expensive calls
are still there.

The changes are undoubtedly an improvement and the patch is fine (apart from some style "violations") but we need to understand what is their impact. My educated guess is: negligible.
But I really hope you will prove me wrong :-)

**Comment by pritha-srivastava** [comment]
@germanop - The VBD.plug call usually takes under 2 seconds to complete and these changes saved a few milliseconds. I am not sure what the impact would have been when I would have tested that for a 16 host bootstorm case (I couldn't because XenRT didnt allow me to)
Can you point me to what calls have been left behind. I have tried to take things that have been passed by XAPI in the command (device_config) and make use of them, but then most of the times I found that the XAPI calls can not be eliminated (for e.g. finding out the master of the pool).

**Comment by germanop** [comment]
In function from_uuid:
* sr_ref = _SR.get_by_uuid(sr_uuid)
* sm_type = _SR.get_type(sr_ref)
* sms = _SM.get_all_records_where('field "type" = "%s"' % sm_type)
* pbds = _PBD.get_all_records_where('field "SR" = "%s" and' % sr_ref +
                                           'field "host" = "%s"' % host_ref)

There are others in the other functions

**Comment by pritha-srivastava** [comment]
@germanop: I have updated the commit message. Let me know if there are other changes to be done.

File: drivers/blktap2.py [hunk 1]
```diff
@@ -1410,18 +1410,20 @@ def call_pluginhandler(cls, session, host_ref, sr_uuid, vdi_uuid, action,
...
+        host_ref = params["host_ref"]
         sm_config = self._session.xenapi.VDI.get_sm_config(vdi_ref)
         attached_as = util.attached_as(sm_config)
         if NO_MULTIPLE_ATTACH and (attached_as == "RW" or \
                 (attached_as == "RO" and attach_mode == "RW")):
             util.SMlog("need to reset VDI %s" % vdi_uuid)
             if not resetvdis.reset_vdi(self._session, vdi_uuid, force=False,
-                    term_output=False, writable=writable):
+                    term_output=False, writable=writable, vdiref=vdi_ref, 
+                    srref=host_ref, smconf=sm_config):
```
Comment (germanop) [inline drivers/blktap2.py]: maybe sr_ref?

File: drivers/resetvdis.py [hunk 2]
```diff
@@ -53,10 +53,23 @@ def reset_sr(session, host_uuid, sr_uuid, is_sr_master):
     sr_lock.release()
     gc_lock.release()
 
-def reset_vdi(session, vdi_uuid, force, term_output=True, writable=True):
-    vdi_ref = session.xenapi.VDI.get_by_uuid(vdi_uuid)
-    vdi_rec = session.xenapi.VDI.get_record(vdi_ref)
-    sm_config = vdi_rec["sm_config"]
+def reset_vdi(session, vdi_uuid, force, term_output=True, writable=True, vdiref=None, srref=None, smconf={}):
```
Comment (germanop) [inline drivers/resetvdis.py]: Too long.

File: drivers/resetvdis.py [hunk 3]
```diff
@@ -53,10 +53,23 @@ def reset_sr(session, host_uuid, sr_uuid, is_sr_master):
...
+    if vdiref is None :
```
Comment (germanop) [inline drivers/resetvdis.py]: no space before colon

File: drivers/resetvdis.py (hunk 2 above)
Comment (germanop) [inline drivers/resetvdis.py]: You can call vdiref -> vdi_ref, srref -> sr_ref, smconf -> sm_conf for better readability and to improve implementation. See below

File: drivers/resetvdis.py (hunk 3 above)
Comment (germanop) [inline drivers/resetvdis.py]: If vdi_Ref is used instead of vdiref, you can skip the "else" block.

File: drivers/resetvdis.py [hunk 4]
```diff
@@ -53,10 +53,23 @@ def reset_sr(session, host_uuid, sr_uuid, is_sr_master):
...
+        vdi_ref = session.xenapi.VDI.get_by_uuid(vdi_uuid)
+    else:
+        vdi_ref = vdiref
+
+    if smconf is None:
```
Comment (germanop) [inline drivers/resetvdis.py]: Same here

File: drivers/resetvdis.py [hunk 5]
```diff
@@ -53,10 +53,23 @@ def reset_sr(session, host_uuid, sr_uuid, is_sr_master):
...
+        vdi_rec = session.xenapi.VDI.get_record(vdi_ref)
+        sm_config = vdi_rec["sm_config"]
+    else:
+        sm_config = smconf
+
+    if srref is None:
```
Comment (germanop) [inline drivers/resetvdis.py]: Same here

File: drivers/blktap2.py [hunk 6]
```diff
@@ -1410,18 +1410,20 @@ def call_pluginhandler(cls, session, host_ref, sr_uuid, vdi_uuid, action,
...
+        vdi_ref = params["vdi_ref"]
+        host_ref = params["host_ref"]
         sm_config = self._session.xenapi.VDI.get_sm_config(vdi_ref)
         attached_as = util.attached_as(sm_config)
         if NO_MULTIPLE_ATTACH and (attached_as == "RW" or \
                 (attached_as == "RO" and attach_mode == "RW")):
             util.SMlog("need to reset VDI %s" % vdi_uuid)
             if not resetvdis.reset_vdi(self._session, vdi_uuid, force=False,
-                    term_output=False, writable=writable):
+                    term_output=False, writable=writable, vdiref=vdi_ref,
```
Comment (germanop) [inline drivers/blktap2.py]: vdiref, srref, smconf names should be changed: see comment at function definition

**Comment by pritha-srivastava** [comment]
@germanop : All the review comments have been incorporated

**Pull request closed without merging**